src/ripple/app/paths/impl/BookStep.cpp

```diff
@@ -237,60 +259,104 @@ void limitStepOut (Quality const& ofrQ,
 */
-template <class TAmtIn, class TAmtOut, class Callback>
-static
+template <class TIn, class TOut>
+template <class Callback>
 std::pair<boost::container::flat_set<uint256>, std::uint32_t>
-forEachOffer (
+BookStep<TIn, TOut>::forEachOffer (
     PaymentSandbox& sb,
     ApplyView& afView,
-    Book const& book,
-    AccountID const& src,
-    AccountID const& dst,
     bool prevStepRedeems,
-    bool ownerPaysTransferFee,
-    Callback& callback,
-    std::uint32_t limit,
-    beast::Journal j)
+    Callback& callback) const
 {
     // Charge the offer owner, not the sender
     // Charge a fee even if the owner is the same as the issuer
     // (the old code does not charge a fee)
     // Calculate amount that goes to the taker and the amount charged the offer owner
-    auto rate = [&](AccountID const& id)->std::uint32_t
+    auto rate = [this, &sb](AccountID const& id)->std::uint32_t
     {
-        if (isXRP (id) || id == dst)
+        if (isXRP (id) || id == this->strandDst_)
             return QUALITY_ONE;
         return transferRate (sb, id).value;
     };
 
     std::uint32_t const trIn = prevStepRedeems
-        ? rate (book.in.account)
+        ? rate (book_.in.account)
         : QUALITY_ONE;
     // Always charge the transfer fee, even if the owner is the issuer
-    std::uint32_t const trOut = ownerPaysTransferFee
-        ? rate (book.out.account)
+    std::uint32_t const trOut = ownerPaysTransferFee_
+        ? rate (book_.out.account)
         : QUALITY_ONE;
 
-    typename FlowOfferStream<TAmtIn, TAmtOut>::StepCounter counter (limit, j);
-    FlowOfferStream<TAmtIn, TAmtOut> offers (
-        sb, afView, book, sb.parentCloseTime (), counter, j);
+    typename FlowOfferStream<TIn, TOut>::StepCounter
+        counter (maxOffersToConsume_, j_);
+
+    FlowOfferStream<TIn, TOut> offers (
+        sb, afView, book_, sb.parentCloseTime (), counter, j_);
 
+    bool offerAttempted = false;
     boost::optional<Quality> ofrQ;
     while (offers.step ())
     {
         auto& offer = offers.tip ();
+
+        // Note that offer.quality() returns a (non-optional) Quality.  So
+        // ofrQ is always safe to use below this point in the loop.
         if (!ofrQ)
             ofrQ = offer.quality ();
         else if (*ofrQ != offer.quality ())
             break;
 
+        // Special rule for offer crossing.  If:
+        //   a. We're offer crossing using default path (no autobridging), and
+        //   b. The offer's quality is at least as good as our quality, and
+        //   c. We're about to cross one of our own offers, then
+        //   c. Delete the old offer.
+        // See explanation in the testSelfCrossOffer1() unit test.
+        if (limitSelfCrossingQuality_ &&
+            *ofrQ >= *limitSelfCrossingQuality_ &&
+            strandSrc_ == offer.owner() &&
+            strandDst_ == offer.owner())
+        {
+            // Remove this offer even if no crossing occurs.
+            offers.permRmOffer (offer.key());
```

Comment:
I agree this logic belongs outside of the `OfferStream`. It's a shame we had to add two new functions to `OfferStream` to support it.